Handle env vars that use valueFrom

jhadvig opened the pull request:
When env vars have speciefied `valueFrom` instead of the `value` we need to render them as well in your `environments` directive.
Adding two proposed solutions that I've liked the most, but still not 100% sure about the solution. Think is that in case of the `configMapKeyRef` & `secretKeyRef` there could be more then 'key' and 'name' based on the https://github.com/kubernetes/kubernetes/blob/b45bf88105159cdbdd112e899bcb073e2b1ae573/pkg/api/types.go#L2048-L2050 so I didnt wanted to have a separate column for each field. On the other hand I wanted to stay by the table concept since the default original envVars are using it.
![screenshot-11](https://cloud.githubusercontent.com/assets/1668218/13678486/561cd842-e6ef-11e5-94aa-321cc7376c6a.png)
![screenshot-12](https://cloud.githubusercontent.com/assets/1668218/13678489/57d35f94-e6ef-11e5-8a27-498540c50e1e.png)
The code is not in the the best shape so for now please skip the code review.

@jwforres @spadgett @ajacobs21e thoughts?

closes https://github.com/openshift/origin/issues/7886

spadgett commented:
If we don't have a consistent set of columns we can use, I don't think a table is right. Maybe just have a referenced variables section with definition lists for each variable?

spadgett commented:
Yeah I was thinking a "Dynamic Variables" section or similar. (I'm OK with a different name there.)

Forgive the ASCII art.

```
 -------------
| Environment |

 -------------
| Name | Val  |
| FOO  | bar  |
| ...  |      |
 -------------

Dynamic Variables

POD_NAMESPACE
Type: ...
Field Path: ...

MY_SECRET_DATA
...
```

jhadvig commented:
@spadgett I've addressed the comments. PTAL
![screenshot-13](https://cloud.githubusercontent.com/assets/1668218/13703072/5dc74db2-e793-11e5-8e88-0d3bc7b335ef.png)

jwforres commented:
we've been moving towards humanizing types everywhere, maybe Field instead
of fieldRef and Secret Key instead of secretKeyRef.  you could do it
generically (so we handle future types) by replacing "Ref" with "" and then
using @spadgett's helper that does the splitting / upper-casing

On Fri, Mar 11, 2016 at 8:13 AM, Jakub Hadvig <notifications@github.com>
wrote:

> @spadgett <https://github.com/spadgett> I've addressed the comments. PTAL
> [image: screenshot-13]
> <https://cloud.githubusercontent.com/assets/1668218/13703072/5dc74db2-e793-11e5-8e88-0d3bc7b335ef.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7931#issuecomment-195357762>.
>

jwforres commented:
I'm wondering if the Container header needs to be bigger, right now it
looks like the Container and the Referenced Variables headers are the same
size, when Ref Vars should be a sub-section of the container.

On Fri, Mar 11, 2016 at 8:30 AM, Jessica Forrester <jforrest@redhat.com>
wrote:

> we've been moving towards humanizing types everywhere, maybe Field instead
> of fieldRef and Secret Key instead of secretKeyRef.  you could do it
> generically (so we handle future types) by replacing "Ref" with "" and then
> using @spadgett's helper that does the splitting / upper-casing
>
> On Fri, Mar 11, 2016 at 8:13 AM, Jakub Hadvig <notifications@github.com>
> wrote:
>
>> @spadgett <https://github.com/spadgett> I've addressed the comments. PTAL
>> [image: screenshot-13]
>> <https://cloud.githubusercontent.com/assets/1668218/13703072/5dc74db2-e793-11e5-8e88-0d3bc7b335ef.png>
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/7931#issuecomment-195357762>.
>>
>
>

jwforres commented:
I can't decide if I like the the hr between the different referenced
variables, i think that's inconsistent with what we do for volumes, volumes
are very similar, they have types and each type has different metadata
associated with it.  I know @rhamilto has been looking at that Volumes
section some so CC'ing here

On Fri, Mar 11, 2016 at 8:31 AM, Jessica Forrester <jforrest@redhat.com>
wrote:

> I'm wondering if the Container header needs to be bigger, right now it
> looks like the Container and the Referenced Variables headers are the same
> size, when Ref Vars should be a sub-section of the container.
>
> On Fri, Mar 11, 2016 at 8:30 AM, Jessica Forrester <jforrest@redhat.com>
> wrote:
>
>> we've been moving towards humanizing types everywhere, maybe Field
>> instead of fieldRef and Secret Key instead of secretKeyRef.  you could do
>> it generically (so we handle future types) by replacing "Ref" with "" and
>> then using @spadgett's helper that does the splitting / upper-casing
>>
>> On Fri, Mar 11, 2016 at 8:13 AM, Jakub Hadvig <notifications@github.com>
>> wrote:
>>
>>> @spadgett <https://github.com/spadgett> I've addressed the comments.
>>> PTAL
>>> [image: screenshot-13]
>>> <https://cloud.githubusercontent.com/assets/1668218/13703072/5dc74db2-e793-11e5-8e88-0d3bc7b335ef.png>
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/openshift/origin/pull/7931#issuecomment-195357762>.
>>>
>>
>>
>

jhadvig commented:
@jwforres +1 for making the container header bigger

jhadvig commented:
Regarding the delimiter I wasnt sure as well, but put it there anyway so I get some feedback. In fact I kinda like this a bit more since you can optically differentiate between the envVars. But yeah otherwise its similar with the volumes

jwforres commented:
So looking at this for a running container... I would really like to know what value that environment reference was resolved to.  AFAIK not possible with the current api response.  @smarterclayton @deads2k is that even technically possible to do later?

jhadvig commented:
Yeah I dont think that is possielbe ATM

rhamilto commented:
> I can't decide if I like the the hr between the different referenced
variables, i think that's inconsistent with what we do for volumes, volumes
are very similar, they have types and each type has different metadata
associated with it.  I know @rhamilto has been looking at that Volumes
section some so CC'ing here

The HR is inconsistent with what we're doing on Volumes [1], and I feel the HR is unnecessary noise.  

[1] https://github.com/openshift/origin/pull/7937.

jhadvig commented:
@rhamilto went throught your PR and I like the styling. Will update accordingly :+1:

deads2k commented:
> So looking at this for a running container... I would really like to know what value that environment reference was resolved to. AFAIK not possible with the current api response. @smarterclayton @deads2k is that even technically possible to do later?

You mean, "can you determine the value you got for `valueFrom.fieldRef.fieldPath`" sort of thing?  Not in a way that will make you happy.  I think only the kubelet and docker know the value that it used and we don't have an API for "get me metadata for this actual container".

You could write a parser that takes a good guess assuming stable-ish values for things you're referencing, but if they've changed that's not necessarily what the kubelet used.

jwforres commented:
On Fri, Mar 11, 2016 at 9:15 AM, David Eads <notifications@github.com>
wrote:

> So looking at this for a running container... I would really like to know
> what value that environment reference was resolved to. AFAIK not possible
> with the current api response. @smarterclayton
> <https://github.com/smarterclayton> @deads2k <https://github.com/deads2k>
> is that even technically possible to do later?
>
> You mean, "can you determine the value you got for
> valueFrom.fieldRef.fieldPath" sort of thing? Not in a way that will make
> you happy. I think only the kubelet and docker know the value that it used
> and we don't have an API for "get me metadata for this actual container".
>

Yeah "get me metadata for this actual container" is really what I was
hoping for, that's a shame :-\

so @jhadvig maybe we have a tip that says something like "( i ) On running
pods you can check what these reference values have resolved to by going to
the Terminal tab and echoing the environment variable."    @spadgett
thoughts?



> You could write a parser that takes a good guess assuming stable-ish
> values for things you're referencing, but if they've changed that's not
> necessarily what the kubelet used.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7931#issuecomment-195382536>.
>

jhadvig commented:
@jwforres @rhamilto I've addressed your comments. PTAL
![screenshot-14](https://cloud.githubusercontent.com/assets/1668218/13708533/68c0a918-e7af-11e5-84a4-640865b0196a.png)
![screenshot-15](https://cloud.githubusercontent.com/assets/1668218/13708535/69a8b78a-e7af-11e5-8e03-2e529f91afc5.png)

smarterclayton commented:
I would expect the layout of these to be in the container table.  This
makes them look special.  Also, it ignores ordering, and ordering is
critical.  The order of the env vars in the pod should *never* be changed
on display.

On Fri, Mar 11, 2016 at 11:34 AM, Jakub Hadvig <notifications@github.com>
wrote:

> @jwforres <https://github.com/jwforres> @rhamilto
> <https://github.com/rhamilto> I've addressed your comments. PTAL
> [image: screenshot-14]
> <https://cloud.githubusercontent.com/assets/1668218/13708533/68c0a918-e7af-11e5-84a4-640865b0196a.png>
> [image: screenshot-15]
> <https://cloud.githubusercontent.com/assets/1668218/13708535/69a8b78a-e7af-11e5-8e03-2e529f91afc5.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7931#issuecomment-195440825>.
>

smarterclayton commented:
Because you can turn around and do env(10) FOO=$(MY_SECRET_DATA) and that
is a different outcome when env(10) is before or after MY_SECRET_DATA's
definition.

On Fri, Mar 11, 2016 at 12:46 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I would expect the layout of these to be in the container table.  This
> makes them look special.  Also, it ignores ordering, and ordering is
> critical.  The order of the env vars in the pod should *never* be changed
> on display.
>
> On Fri, Mar 11, 2016 at 11:34 AM, Jakub Hadvig <notifications@github.com>
> wrote:
>
>> @jwforres <https://github.com/jwforres> @rhamilto
>> <https://github.com/rhamilto> I've addressed your comments. PTAL
>> [image: screenshot-14]
>> <https://cloud.githubusercontent.com/assets/1668218/13708533/68c0a918-e7af-11e5-84a4-640865b0196a.png>
>> [image: screenshot-15]
>> <https://cloud.githubusercontent.com/assets/1668218/13708535/69a8b78a-e7af-11e5-8e03-2e529f91afc5.png>
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/7931#issuecomment-195440825>.
>>
>
>

jwforres commented:
@jhadvig given @smarterclayton's feedback we have to put it back into the table and we need to double check that the table ordering is exactly what we are getting back from the API.

so to distinguish these values in the table as being "special" I was going to propose putting http://fontawesome.io/icon/external-link-square/ in front of the value when they are references

and instead of calling out the various field names do something like

[->] Secret foo, key bar
[->] Field metadata.namespace
etc...

jhadvig commented:
@jwforres I've updated the PR based on our and @smarterclayton comments. Hopefully my styling estimations were accurate. PTAL
![screenshot-19](https://cloud.githubusercontent.com/assets/1668218/13746191/8431f180-e9f2-11e5-8487-20da9da63fbc.png)
![screenshot-20](https://cloud.githubusercontent.com/assets/1668218/13746193/854f99e6-e9f2-11e5-9031-8b1f828a8936.png)

jwforres commented:
I was thinking more like a one line summary of the from information, and don't duplicate the environment name, so for your two examples it would be

```
|  POD_NAMESPACE   |  [->] Field metadata.namespace        |
|  MY_SECRET_DATA  |  [->] Secret test.secret, key data-1  |
```

where | is my ascii art to indicate the columns

jhadvig commented:
@jwforres updated
![screenshot-23](https://cloud.githubusercontent.com/assets/1668218/13753392/0c281234-ea13-11e5-94de-b971d7179c67.png)

smarterclayton commented:
The value column is hard to read, i have no idea what is a user entered
value vs descriptive text.

On Mon, Mar 14, 2016 at 1:32 PM, Jakub Hadvig <notifications@github.com>
wrote:

> @jwforres <https://github.com/jwforres> updated
> [image: screenshot-23]
> <https://cloud.githubusercontent.com/assets/1668218/13753392/0c281234-ea13-11e5-94de-b971d7179c67.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7931#issuecomment-196425269>.
>

jhadvig commented:
@smarterclayton had it like this:

![screenshot-22](https://cloud.githubusercontent.com/assets/1668218/13753793/ca7c7666-ea14-11e5-848c-c4b4a0391091.png)

But we decided with @jwforres and @spadgett to dont distinguish between key and value.

smarterclayton commented:
It's really, really, really hard to parse when it's just text.  This is a
structured field, it should field like it in some form.  I'm not even sure
the current text adequately conveys what each one does.

On Mon, Mar 14, 2016 at 1:45 PM, Jakub Hadvig <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> had it like this:
>
> [image: screenshot-22]
> <https://cloud.githubusercontent.com/assets/1668218/13753793/ca7c7666-ea14-11e5-848c-c4b4a0391091.png>
>
> But we decided with @jwforres <https://github.com/jwforres> and @spadgett
> <https://github.com/spadgett> to dont distinguish between key and value.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7931#issuecomment-196430663>.
>

jhadvig commented:
@smarterclayton @jwforres updated. As suggested I've put there a sentence that would make thing more clear, together with the popover + the sentence uses different font
![screenshot-25](https://cloud.githubusercontent.com/assets/1668218/13784443/92ed6a96-eacf-11e5-897f-5a141f2e4c16.png)

jwforres commented on assets/app/views/directives/environment.html:
```diff
@@ -1,10 +1,33 @@
+              Set to the field <b>{{env.ref.fieldPath}}</b> in the current object.
+            </span>
+            <span ng-switch-default>
+              {{env.type}}
```
Maybe `Set to a reference on a <b>{{env.type}}</b>`

jwforres commented on assets/app/scripts/directives/resources.js:
```diff
@@ -145,6 +145,24 @@ angular.module('openshiftConsole')
         $scope.expanded = {};
+        var newEnvVars = [];
+
+        angular.forEach($scope.envVars, function(envVar) {
```
is this normalization necessary? you are using a switch to handle different cases already in the view.

if you keep it you have to add a link function with a watch for envVars, controller does not get re-run if the data in the binding changes.

jwforres commented on assets/app/styles/_core.less:
```diff
@@ -976,6 +976,16 @@ h1 small.meta, .build-config-summary .meta {
+      color: @btn-default-color;
+    }
+  }
+  .dl-horizontal {
```
shouldnt be needed anymore since you dropped the dl

jwforres commented on assets/app/styles/_core.less:
```diff
@@ -976,6 +976,16 @@ h1 small.meta, .build-config-summary .meta {
       }
     }
   }
+  .action-inline {
```
i don't see anything with the action-inline class being added, is this still needed?

jhadvig commented:
@jwforres I've reverted the normalization, you were right it wants necessary. PTAL

jwforres commented:
:+1:  LGTM

jwforres commented:
[merge]
